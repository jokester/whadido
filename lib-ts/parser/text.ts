/**
 * Parser for text
 */

import {
    Parser, unit, zero, bind,
    filter, or, orMulti, skip, seq2, seq3,
    biased, iterate, iterateN, reiterate,
} from "./combinator";

export {
    Parser, unit, zero, bind,
    filter, or, orMulti, skip, seq2, seq3,
    biased, iterate, iterateN, reiterate,
} from "./combinator";

export const getChar: Parser<string, string> = (input) => {
    if (typeof input === "string" && input.length)
        return [{
            output: input.charAt(0),
            rest: input.substr(1)
        }];
    else
        return [];
};

export const isLetter = (a: string) => a && a.length === 1 && !!/[a-zA-Z]/.exec(a);
export const isDigit = (a: string) => a && a.length === 1 && !!/[0-9]/.exec(a);
export const charLit = (a: string) => filter(getChar, _ => a === _);

export const stringLit = (a: string) => <Parser<string, string>>((input: string) => input.startsWith(a) ? [{ output: a, rest: input.slice(a.length) }] : []);

export const isWhitespace = (a: string) => a === "\r" || a === "\n" || a === " " || a === "\t";

export const whitespace = filter(getChar, isWhitespace);

export const skipWhitespace = <A>(m: Parser<string, A>) => skip(whitespace, m);

export const digit: Parser<string, string> = filter(getChar, isDigit);

export const word = bind<string, string[], string>(
    reiterate(filter(getChar, isLetter)),
    _ => unit<string, string>(_.join("")));

export const num: Parser<string, number> = bind(
    reiterate(digit), (digits) => digits.length ? unit<string, number>(parseInt(digits.join(""))) : zero);

export const beforeWhitespace: <A>(m: Parser<string, A>) => typeof m = <A>(m: Parser<string, A>) => {
    return bind(m, m1 => rest => /* lookahead: return m1 only if rest is empty or starts with whitespace */
        /^[ \r\n\t]|^$/.exec(rest) ? unit(m1)(rest) : zero(rest));
};

export const spaceDelimted = <A>(m: Parser<string, A>) => beforeWhitespace(skipWhitespace(m));

export const rest: Parser<string, string> = (input: string) => [{ output: input, rest: "" }];

export const eof: Parser<string, symbol> = (input: string) => input.length ? [] : [{ output: EOF, rest: undefined }];
export const EOF = Symbol("EOF");
