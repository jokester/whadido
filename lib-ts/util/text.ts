const fs = require('fs');

export function chunkToLines(chunk: string | Buffer) {
    return chunk.toString().split(/\r\n|\r|\n/);
}

export function readFile(filename: string): Promise<string> {
    return new Promise((resolve, reject) => {
        fs.readFile(filename, function (err: any, content: string) {
            if (err)
                reject(err);
            else
                resolve(content);
        })
    });
}

export function readLines(filename: string) {
    return readFile(filename).then(chunkToLines);
}