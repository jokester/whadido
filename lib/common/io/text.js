"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function chunkToLines(chunk) {
    return chunk.toString().split(/\r\n|\r|\n/);
}
exports.chunkToLines = chunkToLines;
//# sourceMappingURL=text.js.map