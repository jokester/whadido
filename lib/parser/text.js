"use strict";
/**
 * Parser for text
 */
Object.defineProperty(exports, "__esModule", { value: true });
const combinator_1 = require("./combinator");
var combinator_2 = require("./combinator");
exports.unit = combinator_2.unit;
exports.zero = combinator_2.zero;
exports.bind = combinator_2.bind;
exports.filter = combinator_2.filter;
exports.or = combinator_2.or;
exports.orMulti = combinator_2.orMulti;
exports.skip = combinator_2.skip;
exports.seq2 = combinator_2.seq2;
exports.seq3 = combinator_2.seq3;
exports.biased = combinator_2.biased;
exports.iterate = combinator_2.iterate;
exports.iterateN = combinator_2.iterateN;
exports.reiterate = combinator_2.reiterate;
exports.getChar = (input) => {
    if (typeof input === "string" && input.length)
        return [{
                output: input.charAt(0),
                rest: input.substr(1)
            }];
    else
        return [];
};
exports.isLetter = (a) => a && a.length === 1 && !!/[a-zA-Z]/.exec(a);
exports.isDigit = (a) => a && a.length === 1 && !!/[0-9]/.exec(a);
exports.charLit = (a) => combinator_1.filter(exports.getChar, _ => a === _);
exports.stringLit = (a) => ((input) => input.startsWith(a) ? [{ output: a, rest: input.slice(a.length) }] : []);
exports.isWhitespace = (a) => a === "\r" || a === "\n" || a === " " || a === "\t";
exports.whitespace = combinator_1.filter(exports.getChar, exports.isWhitespace);
exports.skipWhitespace = (m) => combinator_1.skip(exports.whitespace, m);
exports.digit = combinator_1.filter(exports.getChar, exports.isDigit);
exports.word = combinator_1.bind(combinator_1.reiterate(combinator_1.filter(exports.getChar, exports.isLetter)), _ => combinator_1.unit(_.join("")));
exports.num = combinator_1.bind(combinator_1.reiterate(exports.digit), (digits) => digits.length ? combinator_1.unit(parseInt(digits.join(""))) : combinator_1.zero);
exports.beforeWhitespace = (m) => {
    return combinator_1.bind(m, (m1) => (rest) => /* lookahead: return m1 only if rest is empty or starts with whitespace */ /^[ \r\n\t]|^$/.exec(rest) ? combinator_1.unit(m1)(rest) : combinator_1.zero(rest));
};
exports.spaceDelimted = (m) => exports.beforeWhitespace(exports.skipWhitespace(m));
exports.rest = (input) => [{ output: input, rest: "" }];
exports.eof = (input) => input.length ? [] : [{ output: exports.EOF, rest: undefined }];
exports.EOF = Symbol("EOF");
//# sourceMappingURL=text.js.map