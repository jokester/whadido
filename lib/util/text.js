"use strict";
const fs = require('fs');
function chunkToLines(chunk) {
    return chunk.toString().split(/\r\n|\r|\n/);
}
exports.chunkToLines = chunkToLines;
function readFile(filename) {
    return new Promise((resolve, reject) => {
        fs.readFile(filename, function (err, content) {
            if (err)
                reject(err);
            else
                resolve(content);
        });
    });
}
exports.readFile = readFile;
function readLines(filename) {
    return readFile(filename).then(chunkToLines);
}
exports.readLines = readLines;
//# sourceMappingURL=text.js.map