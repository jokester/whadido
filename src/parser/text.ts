/**
 * Parser for text
 */

import {
  biased,
  bind,
  filter,
  iterate,
  iterateN,
  or,
  orMulti,
  Parser,
  reiterate,
  seq2,
  seq3,
  skip,
  unit,
  zero,
} from './combinator';

export {
  Parser,
  unit,
  zero,
  bind,
  filter,
  or,
  orMulti,
  skip,
  seq2,
  seq3,
  biased,
  iterate,
  iterateN,
  reiterate,
} from './combinator';

export const getChar: Parser<string, string> = input => {
  if (typeof input === 'string' && input.length) {
    return [
      {
        output: input.charAt(0),
        rest: input.substr(1),
      },
    ];
  }
  return [];
};

export const isLetter = (a: string) => (a && a.length === 1 && /[a-zA-Z]/.test(a)) || false;
export const isDigit = (a: string) => (a && a.length === 1 && /[0-9]/.test(a)) || false;
export const charLit = (a: string) => filter(getChar, _ => a === _);

export const stringLit: (a: string) => Parser<string, string> = (a: string) => (input: string) =>
  input.startsWith(a) ? [{ output: a, rest: input.slice(a.length) }] : [];

export const isWhitespace = (a: string) => a === '\r' || a === '\n' || a === ' ' || a === '\t';

export const whitespace = filter(getChar, isWhitespace);

export const skipWhitespace = <A>(m: Parser<string, A>) => skip(whitespace, m);

export const digit: Parser<string, string> = filter(getChar, isDigit);

export const word = bind<string, string[], string>(reiterate(filter(getChar, isLetter)), _ =>
  unit<string, string>(_.join('')),
);

export const num: Parser<string, number> = bind<string, string[], number>(reiterate(digit), digits =>
  digits.length ? unit<string, number>(parseInt(digits.join(''))) : zero,
);

export const beforeWhitespace: <A>(m: Parser<string, A>) => typeof m = <A>(m: Parser<string, A>) => {
  return bind(m, (m1: A) => (rest: string /* lookahead: return m1 only if rest is empty or starts with whitespace */) =>
    /^[ \r\n\t]|^$/.exec(rest) ? unit<string, A>(m1)(rest) : zero(rest),
  );
};

export const spaceDelimited = <A>(m: Parser<string, A>) => beforeWhitespace(skipWhitespace(m));

export const rest: Parser<string, string> = (input: string) => [{ output: input, rest: '' }];

export const eof: Parser<string, symbol> = (input: string) => (input.length ? [] : [{ output: EOF, rest: '' }]);
export const EOF = Symbol('EOF');
