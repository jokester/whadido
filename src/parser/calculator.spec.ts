/**
 * A calculator language from `evaluator` example in baastad paper
 */

import { bind, charLit, num, seq3, skipWhitespace, stringLit, unit } from './text';

namespace cal {
  const grammer = `

Expression := Int
              | '(' Expression ')'
              | Expression '/' Expression

       Int := [1-9][0-9]*

`;

  const TypeAtom = 'atom';
  const TypeDiv = 'div';

  interface Expr {
    type: string;
  }

  interface AtomExpr extends Expr {
    type: typeof TypeAtom;
    val: number;
  }

  interface DivExpr extends Expr {
    type: typeof TypeDiv;
    nom: Expr;
    deno: Expr;
  }

  export const AtomParser = bind(num, _ => unit<string, AtomExpr>({ type: TypeAtom, val: _ }));

  export const ExpParser = skipWhitespace(
    seq3(charLit('('), AtomParser, skipWhitespace(charLit(')')), (a, b, c) => unit<string, AtomExpr>(b)),
  );

  export const DivParser = seq3(AtomParser, stringLit('/'), AtomParser, (a, b, c) =>
    unit<string, DivExpr>({ type: TypeDiv, nom: a, deno: c }),
  );
}

function expectParseResult(parser: (expression: string) => unknown, parserName: string, expression: string) {
  expect(parser(expression)).toMatchSnapshot(`${parserName}(${JSON.stringify(expression)})`);
}

describe('calculator', () => {
  it('parses atom value', () => {
    for (const expression of ['11 ', '  11   ', '11n']) {
      expectParseResult(cal.AtomParser, 'cal.AtomParser', expression);
    }
  });

  it('parses atom value with parenthesis', () => {
    for (const expression of [' ( 111 ) ']) {
      expectParseResult(cal.ExpParser, 'cal.ExpParser', expression);
    }
  });

  it('parses div expression', () => {
    for (const expression of ['111/222', '111   /222', '111/', '/222 ']) {
      expectParseResult(cal.DivParser, 'cal.DivParser', expression);
    }
  });
});
