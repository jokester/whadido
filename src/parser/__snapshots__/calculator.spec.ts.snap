// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator parses atom value with parenthesis: cal.ExpParser(" ( 111 ) ") 1`] = `Array []`;

exports[`calculator parses atom value: cal.AtomParser("  11   ") 1`] = `Array []`;

exports[`calculator parses atom value: cal.AtomParser("11 ") 1`] = `
Array [
  Object {
    "output": Object {
      "type": "atom",
      "val": 11,
    },
    "rest": " ",
  },
]
`;

exports[`calculator parses atom value: cal.AtomParser("11n") 1`] = `
Array [
  Object {
    "output": Object {
      "type": "atom",
      "val": 11,
    },
    "rest": "n",
  },
]
`;

exports[`calculator parses div expression: cal.DivParser("/222 ") 1`] = `Array []`;

exports[`calculator parses div expression: cal.DivParser("111   /222") 1`] = `Array []`;

exports[`calculator parses div expression: cal.DivParser("111/") 1`] = `Array []`;

exports[`calculator parses div expression: cal.DivParser("111/222") 1`] = `
Array [
  Object {
    "output": Object {
      "deno": Object {
        "type": "atom",
        "val": 222,
      },
      "nom": Object {
        "type": "atom",
        "val": 111,
      },
      "type": "div",
    },
    "rest": "",
  },
]
`;
